addons/mail/models/mail_thread.py

```diff
@@ -430,8 +430,11 @@ def _prepare_tracking(self, fields):
         fnames = self._get_tracked_fields().intersection(fields)
         if not fnames:
             return
-        func = self.browse()._finalize_tracking
-        [initial_values] = self.env.cr.precommit.add(func, dict)
+        key = 'mail_tracking.%s' % self._name
+        initial_values = self.env.cr.precommit.data.get(key)
+        if initial_values is None:
+            self.env.cr.precommit.add(self._finalize_tracking)
+            initial_values = self.env.cr.precommit.data[key] = {}
```

Comment:
By the way, if we don't mind duplicating callbacks, things get much simpler:
```suggestion
        self.env.cr.precommit.add(self._finalize_tracking)
        initial_values = self.env.cr.precommit.data.setdefault('mail_tracking.%s' % self._name, {})
```
The callback simply has to clean up the data dict (`pop` with default) to avoid multiple processing.

We could make the deduplication automatic by extending the API of `add` with an optional `key`:
```python
precommit.add(func, key)    # add func if key not in precommit.data
```

Reply:
*(sections re-ordered)*

> xmo: That was the "full-blown" version for replacing the deduplicated-call-with-data of the PR. If you just want to add a callback you can just `hook.add(key, hookfunc)`, you never have to care about what it returns.

> rco: We could make the deduplication automatic by extending the API of add with an optional key (..)

Indeed, and my point is that the "key" and/or the "returned data" you propose are then part of the signature of the main entrypoint for callbacks, and are quirks that all developers would be exposed to (e.g. in the API documentation), even though the vast majority will never need or care about them. And before they can decide they don't care, they have to understand what those things are and why they exists, so we're back to square one: a needlessly quirky API leaking complexity even on the basic use case ;-)


> But deduplicating callbacks was _the entire point_ of the original change. If we don't want to support that we can just rollback the entire thing, there's nothing to add or change.

In reality, I believe "deduplication" was not really a problem we wanted to solve. Our problem was to find a way to store transaction metadata, for the very specific case of mail.thread value tracking.

The way @rco-odoo chose to address it, was to globally introduce the notion of multiple registration of the same callback in the API, with deduplication + returning previous callback arguments, to get a way to accumulate parameters by altering those arguments in-place.

Conceptually mail.thread only needs a way to register a callback, plus some way for that callback to identify all value changes that happened during the transaction. That callback could even be registered many times, and the first one to run would flush the tracking data. Optimizing away the extra callbacks is optional. And yes, it can be done by using the transaction storage as state OR by introducing a "key" concept in the API for deduplication (or possibly other ways)
To me the first option is clearly better, because it keeps the API clean, and leaves the "special optimization" as an exception that can be handled with 2 extra lines of code in the only situations where we actually need it, *iff* we care.

By the way it might be possible to implement the `mail.thread` tracking by overriding `BaseModel.flush()` instead of using callbacks, no? The whole issue of deduplicating calls would then be moot, and the first call would flush it all. But you'd still need to maintain the "tracking data" *somewhere*, wouldn't you?

To me this illustrates that our real use cases drive us towards having two simple, yet versatile tools in our API:
- a simple one-shot callback registration mechanism
- a way to store non-persistent transactional data

... and not towards having a more complicated API that handles deduplication as a special case.